### PR TITLE
CIRCUITPY_* switches for JSON, RE, etc. Doc cleanup

### LINKED ISSUE #1911: Fix documentation sections that still refer to u* modules, like ure and ujson
Some native modules that we've renamed back to their CPython names are still referred to with their `u*` names in the _readthedocs_ documentation, e.g. `circuitpython/docs/library/ujson.rst`.

- Rename these files and also fix the module names mentioned inside them.
- Remove caveat language about these modules not being supported or that they may change.

### LINKED ISSUE #4056: docs: support matrix: doesn't show certain modules, e.g., json
In #4055, @yobmod observed that the built in `json` module doesn't appear in the support matrix or on circuitpython.org's list of supported modules.

This is because it is enabled/disabled by `MICROPY_PY_UJSON` in a **.h** file, rather than by `CIRCUITPY_JSON` in a **.mk** file.

It would be nice if either this was converted to work in the usual way, or **docs/shared_bindings_matrix.py** could learn how to find these kinds of modules too.

I don't have a full list of modules that are affected by this, but looking at `help("modules")` on a full build like Metro M4 Express and comparing it to the support matrix should give a pretty good idea of which modules are affected.

### PR DESCRIPTION
**Pull request by dhalbert**
Fixes #4056
Fixes #1911.

- Add `CIRCUITPY_BINASCII`, `CIRCUITPY_ERRNO`, `CIRCUITPY_JSON`, `CIRCUITPY_RE`,  switches to make these non-`shared-binding` modules be switchable like other features.
- Bring documentation about `json`, `re`, and other forrmer `u*` MicroPython native modules up to date .
- Make "Support Matrix" heading clearer.

### CODE REVIEW

**Comment by dhalbert** [comment]
Confirmed that `raspberrypi` and `esp32-s2` builds still have `json`, etc. included.

**Review by tannewt — COMMENTED** [review]
What do you think about moving them to shared-bindings properly? That'd mean the docs use stubs and would be in circuitpython-stubs too.

**Comment by dhalbert** [comment]
> What do you think about moving them to shared-bindings properly? That'd mean the docs use stubs and would be in circuitpython-stubs too.

I want to do that in the long run:, #1899, #1900, #1901, and will come back to that. I started to do it a while ago but it turned out to be more than a half-day or one-day project. I think I have a branch stashed away. This is a good intermediate step and fixes the sad state of the doc.

File: py/circuitpy_mpconfig.h [hunk 1]
```diff
@@ -195,21 +195,14 @@ typedef long mp_off_t;
...
-#endif
 // Opposite setting is deliberate.
-#define MICROPY_PY_UERRNO_ERRORCODE           (!CIRCUITPY_FULL_BUILD)
-#ifndef MICROPY_PY_URE
-#define MICROPY_PY_URE                        (CIRCUITPY_FULL_BUILD)
-#endif
-#define MICROPY_PY_URE_MATCH_GROUPS           (CIRCUITPY_FULL_BUILD)
-#define MICROPY_PY_URE_MATCH_SPAN_START_END   (CIRCUITPY_FULL_BUILD)
-#define MICROPY_PY_URE_SUB                    (CIRCUITPY_FULL_BUILD)
+#define MICROPY_PY_UERRNO_ERRORCODE           (!CIRCUITPY_RE)
```
Comment (jepler) [inline py/circuitpy_mpconfig.h:202]: Seems weird that this one changed to CIRCUITPY_RE

**Review by jepler — CHANGES_REQUESTED** [review]
This does what I thought it would have, plus some things I would have forgotten.  I did spot one maybe-error, though.  Give it a double check.

File: py/circuitpy_mpconfig.h (hunk 1 above)
Comment (dhalbert) [reply py/circuitpy_mpconfig.h:202]: oops, yes, that's wrong.

**Review by jepler — APPROVED** [review]
Thanks!

**Comment by jepler** [comment]
@dhalbert unfortunately this is failing a build due to lack of space --
```
Build simmel for fr took 25.47s and failed
make: Entering directory '/home/runner/work/circuitpython/circuitpython/ports/nrf'
Use make V=1, make V=2 or set BUILD_VERBOSE similarly in your environment to increase build verbosity.

245912 bytes used, -152 bytes free in flash firmware space out of 245760 bytes (240.0kB).
41592 bytes used, 89480 bytes free in ram for stack and heap out of 131072 bytes (128.0kB).
```
I hadn't seen french be the biggest board before now.  I think there are some modules that can be excluded from simmel to get it to fit again.

**Review by jepler — APPROVED** [review]
If it builds, merge it!  That should give plenty of headroom.

**Pull request merged**